### PR TITLE
Add a map.set() function

### PR DESCRIPTION
**Pull request by Awjin**
See https://github.com/sass/sass/pull/2836
See https://github.com/sass/sass-spec/pull/1561

### CODE REVIEW

File: CHANGELOG.md
Comment (nex3) [inline CHANGELOG.md:9]: Nit: code-format all argument names.

File: CHANGELOG.md
Comment (nex3) [inline CHANGELOG.md:6]: Nit: Use active voice ("Adds to or updates" rather than "Allows adding to or updating").

File: lib/src/functions/map.dart
Comment (nex3) [inline lib/src/functions/map.dart:27]: Please follow the pattern used for other functions by declaring this as a top-level field and then including it by name in the `functions` list.

File: lib/src/functions/map.dart
Comment (nex3) [inline lib/src/functions/map.dart:94]: Please document this.

File: lib/src/functions/map.dart
Comment (nex3) [inline lib/src/functions/map.dart:105]: Once I land https://github.com/sass/dart-sass/pull/1077, you can use `.tryMap()` here to avoid the unnecessary catch.

File: lib/src/functions/map.dart
Comment (nex3) [inline lib/src/functions/map.dart:110]: This is pretty nitty, but creating a separate sublist for every recursion could result in a lot of extra data structure allocations which are relatively expensive. It may be worth writing this in an iterative style instead of recursive to avoid that.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (nex3) [inline CHANGELOG.md:3]: Can you add a link to the sass-lang.com documentation as well? It's not written yet, but when it is it'll be at https://sass-lang.com/documentation/modules/map#set.

File: lib/src/functions/map.dart
Comment (Awjin) [reply lib/src/functions/map.dart]: nice!

File: lib/src/functions/map.dart
Comment (Awjin) [reply lib/src/functions/map.dart]: I had trouble expressing this nicely as iterative, so instead I passed an index into the array to avoid sublisting it over and over. WDYT?

File: lib/src/functions/map.dart
Comment (nex3) [inline lib/src/functions/map.dart:116]:
```suggestion
  var nestedMap = mutableMap[key].tryMap() ?? const SassMap.empty();
```

Using `const` constructors ensures that the same empty map instance is re-used everywhere, which saves allocations.

File: lib/src/functions/map.dart
Comment (nex3) [reply lib/src/functions/map.dart]: I can see why a directly iterative solution would be messy, since you'd need to track an explicit stack of parent maps *and* their associated keys. I think this solution works well enough.

**Review by nex3 — APPROVED** [review]
Please don't merge this until I successfully land sass/dart-sass#1080 on the feature branch.

**Comment by nex3** [comment]
Okay, this should be good to go now once you rebase onto the latest feature.nested-maps.

File: lib/src/functions/map.dart
Comment (Awjin) [reply lib/src/functions/map.dart:116]: Don't we want new empty map instances to be created for each key?

Comment (nex3) [reply lib/src/functions/map.dart:116]: SassScript values are all immutable, so there's no value to creating new instances when re-using an existing instance would do.

Comment (Awjin) [reply lib/src/functions/map.dart:116]: Ah, I see

**Pull request merged**
